### PR TITLE
Provide concept emulation for Ranges-TS concepts

### PR DESCRIPTION
**Pull request by apmccartney**
This pull request provides concept emulation for Ranges-TS bidirectional and random access iterator concepts.

This is still a work in progress, but I generally prefer to iterate with library authors when developing a contribution. 

I believe I've conformed the style guidance in the [contributing guidelines](https://github.com/STEllAR-GROUP/hpx/wiki/HPX-Source-Code-Structure-and-Coding-Standards#coding-standards). I've also attempted to emulate the local coding style for the file I've modified (but I would be surprised if I inferred the conventions perfectly). In either case, I'm happy to make any stylistic corrections.

Regarding the type traits defined in this submission, I was torn as to whether they belonged in the `hpx/trait/is_iterator` header or in dedicated files in the `hpx/util` direction. For the moment, I have opted for the former. Direction in this respect would be welcome.

It appears there is no unit testing for the `is_iterator` header. I'm building and running the complete test suite at present, but am willing to provide unit tests to demonstrate the added functionality in isolation. 

## Proposed Changes

  - [x] Add relevant type traits underlying the [BidirectionalIterator](https://en.cppreference.com/w/cpp/experimental/ranges/iterator/BidirectionalIterator) and [RandomAccessIterator](https://en.cppreference.com/w/cpp/experimental/ranges/iterator/RandomAccessIterator) concepts
    - [x]  `addition_result`
    - [x]  `dereference_result`
    - [x]  `equality_result`
    - [x]  `inequality_result`
    - [x]  `inplace_addition_result`
    - [x]  `inplace_subtraction_result`
    - [x]  `predecrement_result`
    - [x]  `preincrement_result`
    - [x]  `postdecrement_result`
    - [x]  `postincrement_result`
    - [x]  `subscript_result`
    - [x]  `subtraction_result`
  - [x] Add a type trait to verify whether a traversal concept is satisfied
    - [x] `satisfy_traversal_concept`
  - [x] Add specializations to `has_traversal` and `belongs_to_iterator_traversal` integrating the concept check.

## Any background context you want to provide?
See #3641

### CODE REVIEW

**Comment by hkaiser** [comment]
> Regarding the type traits defined in this submission, I was torn as to whether they belonged in the hpx/trait/is_iterator header or in dedicated files in the hpx/util direction. For the moment, I have opted for the former. Direction in this respect would be welcome.

Having the new traits in that header should be good enough. I don't think those are of any use outside of iterator categorization.

> It appears there is no unit testing for the is_iterator header. I'm building and running the complete test suite at present, but am willing to provide unit tests to demonstrate the added functionality in isolation.

Tests would be very welcome, indeed.

Thanks for your work on this!

**Comment by apmccartney** [comment]
> Do we still need the explicit handling of the boost traversal categories with the new traits?

In short, for the BidirectionalIterator and RandomAccessIterator, no, and for weaker iterators, yes.

For BidirectionalIterator and RandomAccessIterator, the interface guarantees are sufficient to robustly distinguish between the two concepts. The boost traversal categories really only provide a mechanism to 'lie', i.e. specify a iterator level that is stronger or weaker than your interface would otherwise suggest (which is probably a footgun). However, the situation with ForwardIterator and InputIterator is less straight forward.

Fundementally, the multiple pass guarantee is what distinguishes the ForwardIterator concept from the InputIterator concept in the ranges TS. To my knowledge, there is no means to deduce this from the iterator interface; we really do need to author annotations of some sort to make a determination. Prior to the C++ 20 standard and its various iterator category requirement reductions, the traversal categories of the boost iterator library are the best way I know of to do so.

**Comment by hkaiser** [comment]
The remaining test failures are unrelated.

**Comment by hkaiser** [comment]
@apmccartney IIUC, a user still needs to specialize the Boost traversal category for their iterators if those are input iterators. Is that correct?

**Comment by apmccartney** [comment]
> A user still needs to specialize the Boost traversal category for their iterators if those are input iterators. Is that correct?

If the user's iterator models forward traversal and nothing stronger, then yes, the user must specialize `boost::iterator::iterator_traversal` on their iterator type in order for hpx algorithms to take advantage of that forward traversal gaurantees.

If the user's iterator models bidirectional traversal or stronger, then no, the user need not take any action for hpx algorithms to take advantage of that traversal (regardless of the iterator's nested `iterator_category` typedef).

**Comment by hkaiser** [comment]
> If the user's iterator models bidirectional traversal or stronger, then no, the user need not take any action for hpx algorithms to take advantage of that traversal (regardless of the iterator's nested iterator_category typedef).

@apmccartney Thanks for this explanation. However, when I use the `reduce_iterators` branch (see #3650), merge your changes there, and edit [this line](https://github.com/STEllAR-GROUP/hpx/pull/3650/files#diff-735859631ebab2dd46b401999296501aR27) to `input_iterator_tag`, I still see compilation problems stating that the used iterator is not at least a forward iterator. I think the iterator used in that test exposes the interface of a random access traversable iterator, so things should compile. Would you mind having a look?

**Comment by apmccartney** [comment]
>However, when I use the reduce_iterators branch (see #3650), merge your changes there, and edit this line to input_iterator_tag, I still see compilation problems stating that the used iterator is not at least a forward iterator.

There's apparently a bug in my implementation. Testing to come should sort that out. =)

> Would you mind having a look?

Absolutely. I'll take a look this evening.

**Comment by apmccartney** [comment]
I reviewed the compilation failure for `reduce_3641.cpp` after merging my branch and making the perscribed modications to the file. After a bit of poking around, I was pleased to find that my implementation was simply incomplete rather than seriously flawed. I've made the corresponding extension and the test is now passing on my locally merged branch. I've explained in some detail in the commit message.

> This commit adds a specialization to the `satisfy_traversal_concept` for
the `boost::forward_traversal_tag` provided by the `boost::iterator`
library.
> 
> Forward traversal cannot be robustly verified via concept emulation due
to the ambiguities between iterfaces specified by the `ForwardIterator`
and `InputIterator` concepts in the C++20 ranges TS. That said, any
class that satisfies the `BidirectionalIterator` concept is certain to
also fullfill the `ForwardIterator` concept.
> 
> The specialization established in this commit leverages that fact to
provide partial concept checking for `ForwardIterator`.

I plan to take some time Saturday afternoon to complete a small set of unit tests. Is there anything else you would like to see as part of this merge request?

**Comment by hkaiser** [comment]
> Is there anything else you would like to see as part of this merge request?

@apmccartney Could you add some comment to the specialization of `satisfy_traversal_concept<Iter, boost::forward_traversal_tag>` (https://github.com/STEllAR-GROUP/hpx/pull/3684/files#diff-ad7521def89116999da36bcdfd269fbdR304) explaining your decision, please? A casual reader might otherwise not understand why it is derived from `bidirectional_concept<Iter>` (frankly, my first thought was: 'This must be a typo!'). Otherwise, all is fine and good to go. Thanks!

File: hpx/traits/is_iterator.hpp [hunk 1]
```diff
@@ -301,6 +301,21 @@ namespace hpx { namespace traits
...
+         *
+         * That said, a type which models the BidirectionalIterator concept
+         * also models a ForwardIterator concept, by definition (and the
+         * interface gaurantees on the BidirectionalIterator concept are
+         * sufficient for robust concept checking). Here we provide a
+         * specialization to capture this case, such that, bidirectional and
+         * random access iterators will be properly recognized as satisfying the
+         * ForwardIterator concept.
+         */
         template <typename Iter>
```
Comment (hkaiser) [inline hpx/traits/is_iterator.hpp]: Thanks!

**Comment by hkaiser** [comment]
@apmccartney Some of our algorithm tests fail as a result of your changes (see https://circleci.com/gh/STEllAR-GROUP/hpx/67315 and https://circleci.com/gh/STEllAR-GROUP/hpx/67327). Would you be able to have a look?

**Comment by apmccartney** [comment]
Yes, absolutely. Looking now. I apologize for the delayed response. Recovering from illness.

File: tests/unit/traits/is_iterator.cpp [hunk 2]
```diff
@@ -0,0 +1,921 @@
...
+  }
+  {
+    using iterator = random_access_traversal_iterator;
+    HPX_TEST_MSG((is_forward_iterator<iterator>::value),
+                 "random access traversal input iterator");
+  }
+  {
+    using iterator =
+      hpx::vector_iterator<int, std::vector<int>>;
+    HPX_TEST_MSG((is_forward_iterator<iterator>::value),
```
Comment (apmccartney) [inline tests/unit/traits/is_iterator.cpp]: If I'm reading the error message correctly, this test should replicate the error. However (at least locally), the test is passing.

Are you aware of any specialization of the `is_forward_iterator` template that might be influencing the larger test?

Comment (hkaiser) [reply tests/unit/traits/is_iterator.cpp]: I think I know what's going wrong. Let me see what I can do.

Comment (hkaiser) [reply tests/unit/traits/is_iterator.cpp]: As it turned out, I didn't know what went wrong ;-) Anyways, I was able to fix the issues. Those were caused by a problem in our `iterator_facade` implementation that simply missed to expose the post-increment operator. Sorry for the trouble.

Comment (hkaiser) [reply tests/unit/traits/is_iterator.cpp]: Update: post-increment was not missing, just broken. All should be fine now *fingers crossed*.

**Comment by hkaiser** [comment]
@msimberg this is good to go. The errors on pycicle are unrelated, afaics.

**Comment by hkaiser** [comment]
@msimberg I would like to merge this asap in order to get #3641 and #3650 out of the way.

**Comment by msimberg** [comment]
@hkaiser almost there... @apmccartney unfortunately we still try to support C++11 (although we've discussed dropping it soonish). That means your `auto`s for return types won't work, see [here](http://cdash.cscs.ch//viewBuildError.php?buildid=42334). Would you mind fixing those?

**Comment by hkaiser** [comment]
> @hkaiser almost there... @apmccartney unfortunately we still try to support C++11 (although we've discussed dropping it soonish). That means your autos for return types won't work, see here. Would you mind fixing those?

I'll fix those. Thanks!

**Comment by hkaiser** [comment]
@msimberg should be fine now, let's see how the tests go.

**Comment by hkaiser** [comment]
@msimberg things look good now, I believe. The C++11 builds have succeeded. The tests fail at runtime because of a configuration error in pycicle… Merging now.

**Comment by hkaiser** [comment]
@apmccartney Congrats to you for your first accepted PR into HPX. Please get in contact with @aserio if you're interested in receiving a STE||AR-Group t-shirt. We give that away for all of our first-time contributors.

**Pull request merged**